include/vcpkg/binarycaching.h

```diff
@@ -196,23 +200,41 @@ namespace vcpkg
 
     struct BinaryCache : ReadOnlyBinaryCache
```

Comment:
When adding threading functionality to the codebase that is not effectively 'do a single function but faster', I think there needs to be a discussion in a comment of 'this is how the threads and communication between them work'.

For instance:

```
// compression and upload of binary cache entries happens on a single 'background' thread, `m_push_thread`
// Thread safety is achieved within the binary cache providers by:
//   1. Only using one thread in the background for this work.
//   2. Forming a queue of work for that thread to consume in `m_actions_to_push`, which maintains its own thread safety
//   3. Sending any replies from the background thread through `m_bg_msg_sink`
//   4. Ensuring any supporting data, such as tool exes, is provided before the background thread is started.
//   5. Ensuring that work is not submitted to the background thread until the corresponding `packages` directory to upload is no longer being actively touched by the foreground thread.
```

Reply:
Does your approach here survive https://github.com/microsoft/vcpkg-tool/pull/1076 ? Hard links mean that we have a lot less certainty on the packages directory being 'hermetic'.

Reply:
https://github.com/microsoft/vcpkg-tool/pull/1076 does not have an impact here. The files "copied" to the installed dir are not later changed by another package. Even if they are overwritten, the hard link in the packages folder would still link to the same original file.

Reply:
It has an impact on windows if https://github.com/microsoft/vcpkg-tool/pull/802 gets merged before this. Because when the packages folder gets compressed by 7z, the next feature gets tested and stuff gets removed from the installed dir and on windows you cant remove a hard link even if the linked file is only opened via another hard link -.-

Reply:
I was finally able to catch this situation that happens in combination of this PR with #802:
<img width="1252" alt="image" src="https://github.com/microsoft/vcpkg-tool/assets/41973254/b0733aea-0675-4370-a656-f0a6d71a61b4">

Reply:
Interestingly the catch seems to only happen with this specific rocksdb.lib file and error nearly only happens with lib files. 

And in general it only happens on the dev drive and not on a normal NTFS drive. Strage 😕